reconcile/utils/membershipsources/app_interface_resolver.py

```diff
@@ -0,0 +1,59 @@
+import base64
+from typing import (
+    Any,
+    Callable,
+    Optional,
+)
+
+from reconcile import queries
+from reconcile.gql_definitions.fragments.membership_source import (
+    AppInterfaceMembershipProviderSourceV1,
+)
+from reconcile.gql_definitions.membershipsources.roles import RoleV1
+from reconcile.gql_definitions.membershipsources.roles import (
+    query as mebershipsource_query,
+)
+from reconcile.utils import gql
+from reconcile.utils.membershipsources.models import (
+    ProviderGroup,
+    RoleBot,
+    RoleMember,
+    RoleUser,
+)
+from reconcile.utils.secret_reader import SecretReader
+
+
+def gql_query_func_for_source(
+    source: AppInterfaceMembershipProviderSourceV1,
+) -> Callable[[str, Optional[Any], bool], dict[str, Any]]:
+    settings = queries.get_secret_reader_settings()
+    secret_reader = SecretReader(settings=settings)
+    username = secret_reader.read_secret(source.username)
+    password = secret_reader.read_secret(source.password)
+    basic_auth_info = base64.b64encode(f"{username}:{password}".encode()).decode()
+    gql_api = gql.get_api_for_server(
+        source.url, f"Basic {basic_auth_info}", None, False
+    )
+    return gql_api.query
```

Comment:
`gql_api` need to call `close` to avoid leak, can refactor this into a class with context manager to do auto cleanup.

Reply:
fixed

Reply:
good point. done

Reply:
we have a helper function `gql_api_builder` to use in test, and what I mean is to add `__enter__` and `__exit__` to `GqlApi` class, so anyone writing new functions to use `get_api_for_server` will auto get `gql_api` object with context management.